lib/src/rules/always_depend_on_packages_you_use.dart

```diff
@@ -0,0 +1,112 @@
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+import '../ast.dart';
+
+const _desc = r'Depend on packages you use.';
+
+const _details = r'''
+
+**DO** depend on packages you use.
+
+When importing a package, *always* add a dependency on it to your pubspec.
+
+Depending explicitly on packages that you use ensures they will always exist
+and allows you to put a dependency constraint on them to guard you against
+breaking changes.
+
+Whether this should be a regular dependency or dev_dependency depends on if it
+is imported from a public file (one under either `lib` or `bin`), or some other
+file.
+
+**BAD:**
+```dart
+import 'package:a/a.dart';
+import 'package:b/b.dart';
+```
+
+```yaml
+dependencies:
+  a: ^1.0.0
+```
+
+**GOOD:**
+```dart
+import 'package:a/a.dart';
+import 'package:b/b.dart';
+```
+
+```yaml
+dependencies:
+  a: ^1.0.0
+  b: ^1.0.0
+```
+
+''';
+
+class AlwaysDependOnPackagesYouUse extends LintRule implements NodeLintRule {
+  AlwaysDependOnPackagesYouUse()
+      : super(
+            name: 'always_depend_on_packages_you_use',
+            description: _desc,
+            details: _details,
+            group: Group.pub);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    // Only lint if we have a pubspec.
+    var package = context.package;
+    if (package is! PubWorkspacePackage) return;
+    var pubspec = package.pubspec;
+    if (pubspec == null) return;
+
+    var visitor = _Visitor(this, context, pubspec,
+        isPublicFile: isInPublicDir(context.currentUnit.unit, context.package));
+    registry.addImportDirective(this, visitor);
+    registry.addExportDirective(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final AlwaysDependOnPackagesYouUse rule;
+  final LinterContext context;
+  final Pubspec pubspec;
+  final bool isPublicFile;
+  late final availableDeps = [
+    if (pubspec.dependencies != null)
+      for (var dep in pubspec.dependencies!)
+        if (dep.name?.text != null) dep.name!.text!,
+    if (!isPublicFile && pubspec.devDependencies != null)
+      for (var dep in pubspec.devDependencies!)
+        if (dep.name?.text != null) dep.name!.text!
+  ];
+
+  _Visitor(this.rule, this.context, this.pubspec, {required this.isPublicFile});
+
+  void _checkDirective(UriBasedDirective node) {
+    // Is it a package: import?
+    var importUriContent = node.uriContent;
+    if (importUriContent == null) return;
+    if (!importUriContent.startsWith('package:')) return;
+
+    try {
+      var importUri = Uri.parse(importUriContent);
+      if (importUri.pathSegments.isEmpty) return;
+      var packageName = importUri.pathSegments.first;
+      if (availableDeps.contains(packageName)) return;
+      rule.reportLint(node.uri);
+    } on FormatException catch (_) {}
+  }
+
+  @override
+  void visitImportDirective(ImportDirective node) => _checkDirective(node);
```

Comment:
@bwilkerson: do you think this would be better done at the compilation unit level?  That way we would only build the list of available deps once per compilation (not repeatedly for each directive).

Notice the relative perf: https://github.com/dart-lang/linter/pull/2659/checks?check_run_id=2622882764#step:5:43

Seems like we can do better...

Reply:
I can try using an iterable instead of a list also

Reply:
> That way we would only build the list of available deps once per compilation (not repeatedly for each directive).

The available deps is cached in a field and is only built once, so there shouldn't be any duplicated effort there.

Reply:
Ok I tried a few things here. The most impactful one by far seemed to be not using `Uri.parse` and doing a bit of hacky munging instead.

I did also move up the creation of this list to the rule itself, and passed that into the visitor. That has the effect I believe of doing it only once per file. I could cache it per package as well, but I don't know how invalidation would work?

I also tried using an iterable and a set to see if they would help but they seemed the same.

I will push up the change to avoid Uri.parse, PTAL.

Reply:
https://github.com/dart-lang/linter/runs/2623499572#step:5:115 just a wee bit better ;)

Reply:
> The available deps is cached in a field and is only built once, so there shouldn't be any duplicated effort there.

Oh true, I didn't need to hoist it. Happy to undo that change if you would like, but this seems a bit better?

Reply:
Either way is fine with me. We can look at caching the list with the `WorkspacePackage` when we move the API.